common/src/main/java/com/microsoft/identity/common/internal/cache/ADALOAuth2TokenCache.java

```diff
@@ -33,65 +63,120 @@ protected void initializeSharedPreferencesFileManager(String fileName) {
+        }
 
+        if(!StringExtensions.isNullOrBlank(cacheItem.getFamilyClientId())){
+            setItem(CacheKey.createCacheKeyForFRT(issuer, clientId, userId), cacheItem);
+        }
 
-/*
-        // new tokens will only be saved into preferred cache location
-        mTokenCacheStore.setItem(CacheKey.createCacheKeyForRTEntry(issuer, resource, clientId, userId),
-                TokenCacheItem.createRegularTokenCacheItem(issuer, resource, clientId, result));
+    }
 
+    private void setItem(String key, ADALTokenCacheItem cacheItem){
 
-        if (result.getIsMultiResourceRefreshToken()) {
-            Log.v(TAG, "Save Multi Resource Refresh token to cache");
-            mTokenCacheStore.setItem(CacheKey.createCacheKeyForMRRT(issuer, clientId, userId),
-                    TokenCacheItem.createMRRTTokenCacheItem(issuer, clientId, result));
+        String json = mGson.toJson(cacheItem);
+        String encrypted = encrypt(json);
+        if (encrypted != null) {
+            mSharedPreferencesFileManager.putString(key, encrypted);
+        } else {
+            Log.e(TAG, "Encrypted output is null");
         }
 
-        // Store separate entries for FRT.
-        if (!StringExtensions.isNullOrBlank(result.getFamilyClientId()) && !StringExtensions.isNullOrBlank(userId)) {
-            Log.v(TAG, "Save Family Refresh token into cache");
-            final TokenCacheItem familyTokenCacheItem = TokenCacheItem.createFRRTTokenCacheItem(issuer, result);
-            mTokenCacheStore.setItem(CacheKey.createCacheKeyForFRT(issuer, result.getFamilyClientId(), userId), familyTokenCacheItem);
+    }
+
+
+    /**
+     * Method that allows to mock StorageHelper class and use custom encryption in UTs.
+     */
+    protected StorageHelper getStorageHelper() {
+        synchronized (LOCK) {
+            if (sHelper == null) {
+                Log.v(TAG, "Started to initialize storage helper");
+                sHelper = new StorageHelper(mContext);
+                Log.v(TAG, "Finished to initialize storage helper");
+            }
+        }
+        return sHelper;
+    }
+
+    private String encrypt(String value) {
+        try {
+            return getStorageHelper().encrypt(value);
+        } catch (GeneralSecurityException | IOException e) {
+            Log.e(TAG, ADALError.ENCRYPTION_ERROR.toString(), e);
         }
-        */
 
+        return null;
     }
 
+    private String decrypt(final String key, final String value) {
+        if (StringExtensions.isNullOrBlank(key)) {
+            throw new IllegalArgumentException("encryption key is null or blank");
+        }
+
+        try {
+            return getStorageHelper().decrypt(value);
+        } catch (GeneralSecurityException | IOException e) {
+            Log.e(TAG, ADALError.DECRYPTION_FAILED.toString(), e);
+            //TODO: Implement remove item in this case... not sure I actually want to do this
```

Comment:
This was done to handle the case where the key gets corrupted, apps get stuck in constant crash loops. Open to suggestions if there is a better way to do this.

Reply:
Can you elaborate on this for me.  When/how would the key become corrupted and why is deleting the token state the way to correct that?

Reply:
Opened: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/issues/40